pallets/pooled-staking/src/pools.rs

```diff
@@ -0,0 +1,393 @@
+    fn hold(candidate: &Candidate<T>, delegator: &Delegator<T>) -> Stake<T>;
+
+    /// Set the amount of shares a delegator have for given candidate.
+    fn set_shares(candidate: &Candidate<T>, delegator: &Delegator<T>, value: Shares<T>);
+    /// Set the total amount of shares all delegators have for given candidate.
+    fn set_shares_supply(candidate: &Candidate<T>, value: Shares<T>);
+    /// Set the total amount of currency staked for given candidate / the value of all shares.
+    fn set_total_staked(candidate: &Candidate<T>, value: Stake<T>);
+    /// Set the amount of currency held for that pool in the delegator account.
+    fn set_hold(candidate: &Candidate<T>, delegator: &Delegator<T>, value: Stake<T>);
+
+    /// Get the initial value of a share in case none exist yet.
+    fn initial_share_value() -> Stake<T>;
+
+    /// Convert an amount of shares to an amount of staked currency for given candidate.
+    /// Returns an error if there are no shares for that candidate.
+    fn shares_to_stake(candidate: &Candidate<T>, shares: Shares<T>) -> Result<Stake<T>, Error<T>> {
+        let total_staked = Self::total_staked(candidate).0;
+        let supply = Self::shares_supply(candidate).0;
+        ensure!(!supply.is_zero(), Error::NoOneIsStaking);
+
+        Ok(Stake(shares.0.mul_div(total_staked, supply)?))
+    }
+
+    /// Convert an amount of shares to an amount of staked currency for given candidate.
+    /// If this candidate have no shares then it uses `initial_share_value` to compute the value.
+    fn shares_to_stake_or_init(
+        candidate: &Candidate<T>,
+        shares: Shares<T>,
+    ) -> Result<Stake<T>, Error<T>> {
+        if Self::total_staked(candidate).0.is_zero() {
+            Ok(Stake(shares.0.err_mul(&Self::initial_share_value().0)?))
+        } else {
+            Self::shares_to_stake(candidate, shares)
+        }
+    }
+
+    /// Convert an amount of staked currency to an amount of shares for given candidate.
+    /// Returns an error if there are no shares for that candidate.
+    fn stake_to_shares(candidate: &Candidate<T>, stake: Stake<T>) -> Result<Shares<T>, Error<T>> {
+        let total_staked = Self::total_staked(candidate).0;
+        let supply = Self::shares_supply(candidate).0;
+        ensure!(!supply.is_zero(), Error::NoOneIsStaking);
+
+        Ok(Shares(stake.0.mul_div(supply, total_staked)?))
+    }
+
+    fn computed_stake(
+        candidate: &Candidate<T>,
+        delegator: &Delegator<T>,
+    ) -> Result<Stake<T>, Error<T>> {
+        let shares = Self::shares(candidate, delegator);
+        if shares.0.is_zero() {
+            return Ok(Stake(Zero::zero()));
+        }
+
+        Self::shares_to_stake(candidate, shares)
+    }
+
+    /// Convert an amount of staked currency to an amount of shares for given candidate.
+    /// If this candidate have no shares then it uses `initial_share_value` to compute the value.
+    fn stake_to_shares_or_init(
```

Comment:
From the name of the function I expected that this would write the default value to storage if it doesn't exist (init = initialize), but it instead returns a default value (initial). How about this name instead?

```suggestion
    fn stake_to_shares_or_default(
```

Reply:
Actually, do we need this function? Any reason why can't `stake_to_shares` handle this case?

Reply:
We want to use this default value only when we create shares. When we want to know how much shares/stake a delegator have, it should not default to this default amount.